src/parquet/arrow/reader.cc

```diff
@@ -244,83 +246,186 @@ FlatColumnReader::Impl::Impl(MemoryPool* pool, const ColumnDescriptor* descr,
+  Status FlatColumnReader::Impl::ReadNonNullableBatch<ArrowType, ParquetType>(     \
+      TypedColumnReader<ParquetType> * reader, int64_t values_to_read,             \
+      int64_t * levels_read) {                                                     \
+    int64_t values_read;                                                           \
+    CType* out_ptr = reinterpret_cast<CType*>(data_buffer_ptr_);                   \
+    PARQUET_CATCH_NOT_OK(*levels_read = reader->ReadBatch(values_to_read, nullptr, \
+                             nullptr, out_ptr + valid_bits_idx_, &values_read));   \
+                                                                                   \
+    valid_bits_idx_ += values_read;                                                \
+                                                                                   \
+    return Status::OK();                                                           \
+  }
+
+NONNULLABLE_BATCH_FAST_PATH(::arrow::Int32Type, Int32Type, int32_t)
+NONNULLABLE_BATCH_FAST_PATH(::arrow::Int64Type, Int64Type, int64_t)
+NONNULLABLE_BATCH_FAST_PATH(::arrow::FloatType, FloatType, float)
+NONNULLABLE_BATCH_FAST_PATH(::arrow::DoubleType, DoubleType, double)
+
 template <>
-void FlatColumnReader::Impl::ReadNonNullableBatch<::arrow::TimestampType, Int96Type>(
-    Int96* values, int64_t values_read) {
+Status FlatColumnReader::Impl::ReadNonNullableBatch<::arrow::TimestampType, Int96Type>(
+    TypedColumnReader<Int96Type>* reader, int64_t values_to_read, int64_t* levels_read) {
+  RETURN_NOT_OK(values_buffer_.Resize(values_to_read * sizeof(Int96Type), false));
+  auto values = reinterpret_cast<Int96*>(values_buffer_.mutable_data());
+  int64_t values_read;
+  PARQUET_CATCH_NOT_OK(*levels_read = reader->ReadBatch(
+                           values_to_read, nullptr, nullptr, values, &values_read));
+
   int64_t* out_ptr = reinterpret_cast<int64_t*>(data_buffer_ptr_);
   for (int64_t i = 0; i < values_read; i++) {
-    out_ptr[i] = impala_timestamp_to_nanoseconds(values[i]);
+    *out_ptr++ = impala_timestamp_to_nanoseconds(values[i]);
   }
   valid_bits_idx_ += values_read;
+
+  return Status::OK();
 }
 
 template <>
-void FlatColumnReader::Impl::ReadNonNullableBatch<::arrow::BooleanType, BooleanType>(
-    bool* values, int64_t values_read) {
+Status FlatColumnReader::Impl::ReadNonNullableBatch<::arrow::BooleanType, BooleanType>(
+    TypedColumnReader<BooleanType>* reader, int64_t values_to_read,
+    int64_t* levels_read) {
+  RETURN_NOT_OK(values_buffer_.Resize(values_to_read * sizeof(bool), false));
+  auto values = reinterpret_cast<bool*>(values_buffer_.mutable_data());
+  int64_t values_read;
+  PARQUET_CATCH_NOT_OK(*levels_read = reader->ReadBatch(
+                           values_to_read, nullptr, nullptr, values, &values_read));
+
   for (int64_t i = 0; i < values_read; i++) {
     if (values[i]) { ::arrow::BitUtil::SetBit(data_buffer_ptr_, valid_bits_idx_); }
```

Comment:
I would change this in a separate PR. For now, this PR should be finished and all new portions at least as well tested as before.